Save preferences JSON using UTF-8 encoding. Bulletproof prefs load.

tfmorris opened the pull request:
Fixes #2543 Fixes #2544 Fixes #2627.

Always use UTF-8 to write JSON because platform default encoding
might not be legal JSON (e.g. ISO 8859-1).

Also be more conservative about keeping backups if we fail to write.

Two open questions:
- Are we using the correct filter on write? Looks like it could be wrong.
- Do we want to make sure that we can deserialize the written file before declaring success? That would have saved us from the user data loss due to the write/read mismatch here.

wetneb commented on main/src/com/google/refine/io/ProjectMetadataUtilities.java:
```diff
@@ -60,25 +61,35 @@
         try {
             ParsingUtilities.defaultWriter.writeValue(writer, projectMeta);
+            // TODO: Do we want the below instead? Doesn't filter recon candidates
+//            ParsingUtilities.saveWriter.writeValue(writer, projectMeta);
```
There is no reason for recon candidates to end up in the project metadata, so both writers are equivalent. Perhaps we could use `saveWriter` because we are *saving* things to disk, but it is a style change.

tfmorris replied:
I switched to `saveWriter`

wetneb commented on main/src/com/google/refine/io/ProjectMetadataUtilities.java:
```diff
@@ -60,25 +61,35 @@
     
     protected static void saveToFile(ProjectMetadata projectMeta, File metadataFile) throws IOException   {
-        Writer writer = new OutputStreamWriter(new FileOutputStream(metadataFile));
+        Writer writer = new OutputStreamWriter(new FileOutputStream(metadataFile), StandardCharsets.UTF_8);
```
Perhaps we could look into making the deserializer more flexible with respect to encoding issues (just like it can be configured to accept unescaped tabs)? I have not checked if it is possible though.

tfmorris replied:
Jackson does it's own encoding guessing, but it's got to be one of the JSON legal encodings (UTF-8, 16, 32). The ASCII is legal UTF-8, but the upper half of ISO 8859-1 isn't, which is where, I suspect, the problem comes in.

wetneb commented:
Thanks a lot for looking into this. I am a bit puzzled by this - on top of my mind I don't think we enforce UTF-8 anywhere else, so I would have expected this bug to show up earlier (on project data, changes, any other JSON data that we serialize to disk). Perhaps we use other abstractions which enforce UTF-8 themselves? It would be worth reviewing that.

> Do we want to make sure that we can deserialize the written file before declaring success? That would have saved us from the user data loss due to the write/read mismatch here.

If we do it here, I guess we should do it everywhere where data loss could be an issue… It sounds pretty uninviting!

tfmorris commented:
>  I am a bit puzzled by this - on top of my mind I don't think we enforce UTF-8 anywhere else, so I would have expected this bug to show up earlier (on project data, changes, any other JSON data that we serialize to disk). Perhaps we use other abstractions which enforce UTF-8 themselves? It would be worth reviewing that.

It's really never safe to depend on the default platform encoding, so we should definitely review to make sure there aren't other places where we do that, but since we haven't seen other issues, I suspect we're OK.

The reason that the problem likely wasn't more prevalent is that it's a) Windows only and b) requires non-ASCII data in the preferences to trigger it.  Hmmm, speaking of which, I guess I should write a test case for it too.

wetneb commented:
Yes but I would expect we serialize non-ASCII data in many places in the data grid, for instance, and I never heard of project data being corrupted on Windows…

tfmorris commented:
OK, I think this is ready for review @wetneb.  I switched all the tests to run with cp1252 encoding (the old Windows default) to catch problems like this in the future and wrote a test that I confirmed fails with this setup against `master`.

Good thing I added the test because I found that the problem I fixed wasn't actually the one causing the issue, so I fixed the second bit of code and did a non-exhaustive review of other places where JSON serializations are saved to file and all the rest appear to have the correct UTF-8 encoding set.

I did **not** go the extra mile of making sure that we can re-read the `workspace.json` after we write it out.

In terms of severity, I think this will affect any Windows user who has non-ASCII data in any of:
- expression history (column names, replacement strings, etc)
- preferences
- reconciliation service metadata (e.g. GND reconciliation service)

wetneb commented on pom.xml:
```diff
@@ -57,7 +57,7 @@
     <refine.data>/tmp/refine</refine.data>
     <surefire.version>2.22.2</surefire.version>
-    <surefireArgs></surefireArgs>
+    <surefireArgs>-Dfile.encoding=cp1252</surefireArgs>
```
I guess we should ideally be able to apply this only for specific tests, no? I am not sure how that could be done, though. I am trying to think about the risks we are taking by switching to this encoding everywhere. Intuitively it should help us surface more bugs rather than hide ones which would only happen if the system encoding is UTF-8, so it is probably fine?

tfmorris replied:
That was my thinking. We assume/need UTF-8 generally, but our most popular platform doesn't use that as its default encoding, so neither should our tests. I think all tests should be run with this setting.
